assets/app/view/game/buy_sell_shares.rb

```diff
@@ -12,114 +13,138 @@ class BuySellShares < Snabberb::Component
-            reduced_price = @game.format_currency(pool_share.price - swap_share.price)
-            children << h(
-              :button,
-              { on: { click: -> { buy_share(current_entity, pool_share, swap: swap_share) } } },
-              "Buy #{pool_share.percent}% Market Share (#{reduced_price} + #{swap_share.percent}% Share)",
-            )
-          end
+        children
+      end
+
+      # Put up one buy button for each buyable percentage share type in market.
+      # In case there are more than one type of percentages in market or if shares are not the
+      # standard percent (e.g. 5% in 18MEX), show percentage type on button.
+      # Do skip president's share in case there are other shares available.
+      def render_market_shares
+        @pool_shares.map do |share|
+          next unless @step.can_buy?(@current_entity, share)
+          next if share.president && @pool_shares.size > 1
+
+          h(Button::BuyShare,
+            share: share,
+            entity: @current_entity,
+            percentages_available: @pool_shares.size)
         end
+      end
 
-        if step.current_actions.include?('short') && step.can_short?(current_entity, @corporation)
-          short = lambda do
-            process_action(Engine::Action::Short.new(current_entity, corporation: @corporation))
-          end
+      def render_ipo_shares
+        @ipo_shares.map do |share|
+          next unless @step.can_buy?(@current_entity, share)
 
-          children << h(
-            :button,
-            { on: { click: short } },
-            'Short Share',
-          )
+          h(Button::BuyShare,
+            share: share,
+            entity: @current_entity,
+            percentages_available: @ipo_shares.size,
+            source: @game.class::IPO_NAME)
         end
+      end
 
-        # Allow privates to be exchanged for shares
-        @game.companies.each do |company|
-          company.abilities(:exchange) do |ability|
-            next unless ability.corporation == @corporation.name
-            next unless company.owner == current_entity
+      def render_price_protection
+        return [] unless @step.respond_to?(:price_protection)
 
-            prefix = "Exchange #{company.sym} for "
+        price_protection = @step.price_protection
 
-            if ability.from.include?(:ipo) && step.can_gain?(company.owner, ipo_share, exchange: true)
-              children << h(:button, { on: { click: -> { buy_share(company, ipo_share) } } },
-                            "#{prefix} an #{@game.class::IPO_NAME} share")
-            end
+        return [] unless price_protection
+
+        protect = lambda do
+          process_action(Engine::Action::BuyShares.new(@current_entity, shares: price_protection.shares))
+        end
 
-            next unless ability.from.include?(:market)
+        [h(:button, { on: { click: -> { protect } } }, 'Protect Shares')]
```

Comment:
I checked the price_protection as you asked and it's broken, but only because you have a lambda inside another lambda on the click. To keep the style similar to the rest, I suggest you put:

`protect = -> { process_action(Engine::Action::BuyShares.new(@current_entity, shares: price_protection.shares)) }

[h(:button, { on: { click: protect } }, 'Protect Shares')]`

Reply:
Oh wow, I didn't notice the -> at all. I thought I was following what was done for shorts but a couple extra characters makes a difference!

Reply:
Alright, both lambdas / buttons use the same syntax now. I verified Short Sale still works.